// navigation/lib/config/State.js
"use strict";
exports.__esModule = true;
var State = (function () {
    function State() {
        this.defaults = {};
        this.defaultTypes = {};
        this.formattedDefaults = {};
        this.formattedArrayDefaults = {};
        this.trackCrumbTrail = false;
        this.trackTypes = true;
    }
    State.prototype.unloading = function (state, data, url, unload, history) {
        unload();
    };
    ;
    State.prototype.navigating = function (data, url, navigate, history) {
        navigate();
    };
    ;
    State.prototype.dispose = function () {
    };
    ;
    State.prototype.navigated = function (data, asyncData) {
    };
    ;
    State.prototype.urlEncode = function (state, key, val, queryString) {
        return encodeURIComponent(val);
    };
    State.prototype.urlDecode = function (state, key, val, queryString) {
        return decodeURIComponent(val);
    };
    State.prototype.validate = function (data) {
        return true;
    };
    State.prototype.truncateCrumbTrail = function (state, data, crumbs) {
        return crumbs;
    };
    return State;
}());
exports["default"] = State;
