// navigation/lib/StateContext.js
"use strict";
exports.__esModule = true;
var StateContext = (function () {
    function StateContext() {
        this.oldState = null;
        this.oldData = {};
        this.oldUrl = null;
        this.previousState = null;
        this.previousData = {};
        this.previousUrl = null;
        this.state = null;
        this.data = {};
        this.url = null;
        this.title = null;
        this.crumbs = [];
        this.nextCrumb = null;
    }
    StateContext.prototype.clear = function () {
        this.oldState = null;
        this.oldData = {};
        this.oldUrl = null;
        this.previousState = null;
        this.previousData = {};
        this.previousUrl = null;
        this.state = null;
        this.data = {};
        this.url = null;
        this.title = null;
        this.crumbs = [];
        this.nextCrumb = null;
    };
    StateContext.prototype.includeCurrentData = function (data, keys) {
        if (!keys) {
            keys = [];
            for (var key in this.data)
                keys.push(key);
        }
        var newData = {};
        for (var i = 0; i < keys.length; i++)
            newData[keys[i]] = this.data[keys[i]];
        for (var key in data)
            newData[key] = data[key];
        return newData;
    };
    return StateContext;
}());
exports["default"] = StateContext;
